### ISSUE TITLE
[Security] Strange behaviour of Symfony\Component\Security\Http\Firewall\LogoutListener

### ISSUE THREAD

**Issue by xkobal** [symfony/symfony#7104]
Hi,

There is a problem with handler usage in Symfony\Component\Security\Http\Firewall\LogoutListener.

The problem occurs in this case:
- You hit directly the logout route without being authenticated (route defined with roles "IS_AUTHENTICATED_ANONYMOUSLY")
- You use RememberMe

In this case, le first logout won't work because the handlers are only called when a Security Token is set and the "delete RememberMe" handler "Symfony\Component\Security\Http\RememberMe\PersistentTokenBasedRememberMeServices" is one of these handler.
It is strange to think that a logoutAction doesn't always delete RememberMe, meaning that the user will be authenticated at the next hit.

The problem is the same for this handler "Symfony\Component\Security\Http\Logout\CookieClearingLogoutHandler".

Seems that there is no need to restrict the handlers execution with the existence of the SecurityToken because they don't use it:
- PersistentTokenBasedRememberMeServices use only Request
- CookieClearingLogoutHandler use only Response
- And even SessionLogoutHandler use only Request

I think there is three ways for solve that:
- Ensure that the user will be authenticated before the LogoutListener->handle is called
- Remove the Security Token restriction to call these handlers
- Add an INTERACTIVE_LOGOUT event to call these handler (thanks to @alexandresalome for this idea)

If you have any question on that problem or if you need any help (code example) to solve it, I'm in.

Thanks for your help.

Xavier

**Comment by WouterJ** [comment]
I can confirm this bug and I think we should solve it. It indeed is quite strange that a user is not logged out when he accesses the logout url (because he wasn't probably authenticated).

Which of the proposed solutions should be implemented?

**Comment by MacDada** [comment]
Copy-paste from my issue. If any kind of solution is decided, I can implement it.

----

IMHO `LogoutListener` should be called last, after other listeners, which might authenticate user before he/she is logged out. Otherwise none of the handlers are called.

In my situation it is crucial that `RememberMeServices` (as `LogoutHandlerInterface`) + `CookieClearingLogoutHandler` get called, because Symfony co–exists with non–Symfony app.

----

I don't know what's the correct solution, but as a fast hack it could work to:
* override `LogoutListener` and call `LogoutHandlerInterface`s manually with a dummy `TokenInterface` if there is no token yet,
* override `LogoutListener` and do custom actions here (but I believe `LogoutHandlerInterface`s should be used, as they might contain important logic),
* override [`LogoutSuccessHandlerInterface`](https://github.com/symfony/symfony/blob/2.6/src/Symfony/Component/Security/Http/Logout/LogoutSuccessHandlerInterface.php), but same problem as above,
* override `FirewallMap` to change the order of listeners (as it should have already been done in `SecurityExtension`?).

**Comment by morille** [comment]
If one logs-in to a symfony application on a public computer and forget to uncheck the "remember-me"  checkbox, he has currently no way to log-out, except by clearing browser caches.
He may even think that the logout was successfull depending on the redirect.
The next user only has to hit the backspace key.

For people using the "remember me" functionnality you HAVE TO delete the cookie from the logout success handler (which is always called) or you have a big security issue.

**Comment by sergeyfedotov** [comment]
The same issue occurred for a stateless firewall with a remember_me functionality. Context listener doesn't exists in the stateless firewall and authentication listeners are executed after the logout listener. The simplest solution is to override the order of the listeners. I can't see any drawbacks.

**Comment by stof** [comment]
Moving the LogoutListener at the end is broken, because if any other listener sets a Response in the event the LogoutListener won't be called at all.

We need to find a way to fix the case of the remember_me though

**Comment by MacDada** [comment]
I'm trying to write a functional test of this issue in my app. Any idea how to reproduce it? I want to clear "session" / "session cookie" / "session storage" in a way, that I would have "remember me" cookie, but no session.

```php
public function testLogout(Client $client)
{
	//$client->getKernel()->boot();
	//$client->getContainer()->get('session.storage.filesystem')->clear();
	$client->getRequest()->getSession()->invalidate();
	$client->getCookieJar()->expire('MOCKSESSID', '/', self::DOMAIN);

	$client->request('GET', self::LOGOUT_URI);
```

As the result, I'm successfully logged out. Because of the bug, I shouldn't be ;)

The test gives me `RememberMeToken` in `LogoutListener`, while dev/prod it is `null`, because `RememberMeListener` had no time to create the token yet: https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Security/Http/Firewall/LogoutListener.php#L125

**Comment by MatTheCat** [comment]
I just stumbled upon this and I really would like it fixed. Is it really necessary to pass the token to the `logout` function? Neither `CookieClearingLogoutHandler` nor `SessionLogoutHandler` are using it. Meanwhile the `LogoutListener` is broken.

@stof did you witness any use case that required a token when logging out?

**Comment by ebuildy** [comment]
Woa, cannot believe this bug is still here with Symfony 4! I have just copy/paster security.yaml from official documentation, and the bug is here, cannot logout:

https://stackoverflow.com/questions/47932614/symfony-security-logout-not-clearing-rememberme-token

The `LogoutListener` is called **before** the `RememberMeListener`, hence, there is no token to clean :/

**Comment by ChrisWesterfield** [comment]
I can confirm this as well.
BBug is still Existing!

**Comment by MatTheCat** [comment]
@db306 yes see https://github.com/symfony/symfony/commit/205b097802814cfbfeaef887c4896db0067d8fb5

**Comment by MPLLC** [comment]
Any chance this fix can be applied to 3.4 as well?
